In [1]:
import numpy as np
import pandas as pd
from datetime import datetime

### Read CSV


In [2]:
import numpy as np
import pandas as pd
from datetime import datetime

amazon_data_set = pd.read_csv("C:\\Users\\jyost\\OneDrive\\Desktop\\Overall Info\\Imp\\Fall 2019\\Project\\Amazon Fine Food\\Data set\\Reviews.csv")
amazon_data_set.head()

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3   4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4   5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   

   HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                     1                       1      5  1303862400   
1                     0                       0      1  1346976000   
2                     1                       1      4  1219017600   
3                     3                       3      2  1307923200   
4                     0                       0      5  1350777600   

                 Summary                                               Text  
0  Good Quality Dog Food  I have bought several of the Vitality canned d...  
1      Not as Advertised  Product arrived labeled as Jumbo Salted Peanut...  
2  "Delight" says it all  This is a confection that has been around a fe...  
3         Cough Medicine  If you are looking for the secret ingredient i...  
4            Great taffy  Great taffy at a great price.  There was a wid...

### Removing the Duplicate data


In [3]:
# To identify the number of duplicate rows in the data set 
# amazon_data_set.duplicated(subset = ['ProductId','UserId','Score','Time']).sum()

# To delete the duplicate rows in the data set 
amazon_data_set.drop_duplicates(subset = ['ProductId','UserId','Score','Time']).shape


(565125, 10)

### Converting the data type of Time from int to datetime


In [4]:
amazon_data_set['Time'] = pd.to_datetime(amazon_data_set['Time'],unit='s')
amazon_data_set.head()

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3   4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4   5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   

   HelpfulnessNumerator  HelpfulnessDenominator  Score       Time  \
0                     1                       1      5 2011-04-27   
1                     0                       0      1 2012-09-07   
2                     1                       1      4 2008-08-18   
3                     3                       3      2 2011-06-13   
4                     0                       0      5 2012-10-21   

                 Summary                                               Text  
0  Good Quality Dog Food  I have bought several of the Vitality canned d...  
1      Not as Advertised  Product arrived labeled as Jumbo Salted Peanut...  
2  "Delight" says it all  This is a confection that has been around a fe...  
3         Cough Medicine  If you are looking for the secret ingredient i...  
4            Great taffy  Great taffy at a great price.  There was a wid...

### Adding the Positivity Column 
##### when score >=3 then Positivity = 1, else 0

In [5]:
amazon_data_set.dropna(inplace=True)
amazon_data_set['Positivity'] = np.where(amazon_data_set['Score'] >= 3, 1, 0)
amazon_data_set.head()

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3   4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4   5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   

   HelpfulnessNumerator  HelpfulnessDenominator  Score       Time  \
0                     1                       1      5 2011-04-27   
1                     0                       0      1 2012-09-07   
2                     1                       1      4 2008-08-18   
3                     3                       3      2 2011-06-13   
4                     0                       0      5 2012-10-21   

                 Summary                                               Text  \
0  Good Quality Dog Food  I have bought several of the Vitality canned d...   
1      Not as Advertised  Product arrived labeled as Jumbo Salted Peanut...   
2  "Delight" says it all  This is a confection that has been around a fe...   
3         Cough Medicine  If you are looking for the secret ingredient i...   
4            Great taffy  Great taffy at a great price.  There was a wid...   

   Positivity  
0           1  
1           0  
2           1  
3           0  
4           1

### Calc the mean score for every product and sort the results in descending order

In [6]:
amazon_data_set.groupby('ProductId')['Score'].mean().head()
amazon_data_set.groupby('ProductId')['Score'].mean().sort_values(ascending=False)

ProductId
B009WVB40S    5.0
B0022N49M8    5.0
B0022N89QK    5.0
B0022N6OBM    5.0
B0022N6CH8    5.0
             ... 
B0055X2Y4M    1.0
B000MS78K0    1.0
B00032CT4A    1.0
B002Z3UBPO    1.0
B002UZYNHO    1.0
Name: Score, Length: 74258, dtype: float64

### Calc the number of reviews for every product and sorted the scores in desc order


In [7]:
amazon_data_set.groupby('ProductId')['Score'].count().sort_values(ascending=False)

ProductId
B007JFMH8M    913
B002QWP8H0    632
B002QWHJOU    632
B0026RQTGE    632
B002QWP89S    632
             ... 
B0031YHA5A      1
B0031YHA3M      1
B0031TX6ZI      1
B0031TPDQS      1
141278509X      1
Name: Score, Length: 74258, dtype: int64

### Creating a data frame "product_mean_count" that has ProductId, Mean_Score and Score_counts

In [8]:
#Adding to the mean score to the dataframe
product_mean_count = pd.DataFrame(amazon_data_set.groupby('ProductId')['Score'].mean())
product_mean_count

Score
ProductId           
141278509X  5.000000
2734888454  3.500000
2841233731  5.000000
6641040     4.351351
7310172001  4.751445
...              ...
B009UOFTUI  1.000000
B009UOFU20  1.000000
B009UUS05I  5.000000
B009WSNWC4  5.000000
B009WVB40S  5.000000

[74258 rows x 1 columns]

In [9]:
#Adding to the score count to the dataframe
product_mean_count['score_counts'] = pd.DataFrame(amazon_data_set.groupby('ProductId')['Score'].count())
product_mean_count

Score  score_counts
ProductId                         
141278509X  5.000000             1
2734888454  3.500000             2
2841233731  5.000000             1
6641040     4.351351            37
7310172001  4.751445           173
...              ...           ...
B009UOFTUI  1.000000             1
B009UOFU20  1.000000             1
B009UUS05I  5.000000             1
B009WSNWC4  5.000000             1
B009WVB40S  5.000000             1

[74258 rows x 2 columns]

In [10]:
#Sorting the values
sorted_data = product_mean_count.sort_values('score_counts',ascending=False)
sorted_data

Score  score_counts
ProductId                         
B007JFMH8M  4.582694           913
B002QWP89S  4.588608           632
B002QWP8H0  4.588608           632
B002QWHJOU  4.588608           632
B0026RQTGE  4.588608           632
...              ...           ...
B0029TATL0  5.000000             1
B0029TJR7M  5.000000             1
B0029U7OBM  1.000000             1
B0029USWFE  4.000000             1
B009WVB40S  5.000000             1

[74258 rows x 2 columns]

### Sort the product_mean_count in desc order of Score


In [11]:
#Renaming the Score to mean score
sorted_data.rename(columns={'Score': 'Mean_Score'}, inplace=True)
sorted_data

Mean_Score  score_counts
ProductId                           
B007JFMH8M    4.582694           913
B002QWP89S    4.588608           632
B002QWP8H0    4.588608           632
B002QWHJOU    4.588608           632
B0026RQTGE    4.588608           632
...                ...           ...
B0029TATL0    5.000000             1
B0029TJR7M    5.000000             1
B0029U7OBM    1.000000             1
B0029USWFE    4.000000             1
B009WVB40S    5.000000             1

[74258 rows x 2 columns]

### Calc the percentage_score
##### percentage_score = num of reviews per prod/ Total # of reviews for all products

In [12]:
#Calculating what percent each score count is of the total
grouped_sum = sorted_data['score_counts'].sum()
print (grouped_sum)

sorted_data['percentage']  = sorted_data['score_counts'].div(grouped_sum)*100
product_mean_count_percent = sorted_data.sort_values(['score_counts', 'ProductId'], ascending = [0,1])
print(product_mean_count_percent)



568411
            Mean_Score  score_counts  percentage
ProductId                                       
B007JFMH8M    4.582694           913    0.160623
B0026RQTGE    4.588608           632    0.111187
B002QWHJOU    4.588608           632    0.111187
B002QWP89S    4.588608           632    0.111187
B002QWP8H0    4.588608           632    0.111187
...                ...           ...         ...
B009UOFTUI    1.000000             1    0.000176
B009UOFU20    1.000000             1    0.000176
B009UUS05I    5.000000             1    0.000176
B009WSNWC4    5.000000             1    0.000176
B009WVB40S    5.000000             1    0.000176

[74258 rows x 3 columns]


### Merging the main data set with the product_mean_count_percent data frame on Product Id


In [13]:
merged_dataset = pd.merge(amazon_data_set, product_mean_count_percent, on='ProductId')
len(merged_dataset)
merged_dataset.head()

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3   4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4   5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   

   HelpfulnessNumerator  HelpfulnessDenominator  Score       Time  \
0                     1                       1      5 2011-04-27   
1                     0                       0      1 2012-09-07   
2                     1                       1      4 2008-08-18   
3                     3                       3      2 2011-06-13   
4                     0                       0      5 2012-10-21   

                 Summary                                               Text  \
0  Good Quality Dog Food  I have bought several of the Vitality canned d...   
1      Not as Advertised  Product arrived labeled as Jumbo Salted Peanut...   
2  "Delight" says it all  This is a confection that has been around a fe...   
3         Cough Medicine  If you are looking for the secret ingredient i...   
4            Great taffy  Great taffy at a great price.  There was a wid...   

   Positivity  Mean_Score  score_counts  percentage  
0           1        5.00             1    0.000176  
1           0        1.00             1    0.000176  
2           1        4.00             1    0.000176  
3           0        2.00             1    0.000176  
4           1        4.75             4    0.000704

### Mimimizing the data set to reduce the noise i.e. removing the products that have less than 5% of the total ratings

In [14]:
#decreasing the dataset size to run
minimized_data_set = merged_dataset[(merged_dataset.Mean_Score>1) & (merged_dataset.percentage>0.05)]
len(minimized_data_set)

63242

## Text Analysis

### Splitting data into random training and test subsets using “Text” and “Positivity” columns

In [15]:
from sklearn.model_selection import train_test_split

In [16]:
X_train, X_test, y_train, y_test = train_test_split(minimized_data_set['Text'], minimized_data_set['Positivity'], test_size = 0.2, random_state = 50)

### Applying CountVectorizer approach
##### CountVectorizer: Number of times a word occurs in a specific document 

In [17]:
from sklearn.feature_extraction.text import CountVectorizer

#Instantiating CountVectorizer and fit into train data

vect = CountVectorizer().fit(X_train)
vect

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)

### Get the words at a step size of 2000

In [18]:
vect.get_feature_names()[::2000]


['00',
 'ash',
 'bubbled',
 'conveni8ent',
 'drew',
 'fortunatly',
 'husk',
 'littles',
 'nuked',
 'practitioners',
 'rooibos',
 'springer',
 'tracked',
 'written']

In [19]:
len(vect.get_feature_names())

26280

In [20]:
X_train_vectorized = vect.transform(X_train)
X_train_vectorized

<50593x26280 sparse matrix of type '<class 'numpy.int64'>'
	with 2882508 stored elements in Compressed Sparse Row format>

### Converting the vector to array
##### Row represents the document and the column represents the words 

In [21]:
X_train_vectorized.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

### Applying Logistic Regression


In [22]:
from sklearn.linear_model import LogisticRegression


In [23]:
# Create a instance for the LogisticRegression method and fit the training data
model1 = LogisticRegression()
model1.fit(X_train_vectorized, y_train)


C:\Users\jyost\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [24]:
from sklearn.metrics import roc_auc_score

predictions = model1.predict(vect.transform(X_test))
print('AUC: ', roc_auc_score(y_test, predictions))

AUC:  0.9119522202258008


###### The result is not bad. 
##### In order to better understand how our model makes these predictions, we can use the coefficients for each feature (a word) to determine its weight in terms of positivity and negativity.

In [25]:
feature_names = np.array(vect.get_feature_names())
sorted_coef_index = model1.coef_[0].argsort()
print('Smallest Coefs: \n{}\n'.format(feature_names[sorted_coef_index[:10]]))
print('Largest Coefs: \n{}\n'.format(feature_names[sorted_coef_index[:-11:-1]]))

Smallest Coefs: 
['crunchie' 'undrinkable' 'stapled' 'yuck' 'caribou' 'gross' 'worst'
 'defeats' 'bleh' 'steve']

Largest Coefs: 
['fantastic' 'crunchy' 'hooked' 'tend' 'smooth' 'delicious' 'complaints'
 'stocking' 'amazing' 'skeptical']



### Tf–idf term weighting

In [26]:
from sklearn.feature_extraction.text import TfidfVectorizer

##### Get rid of the words that occurs in less than 5 documents

In [27]:
vect = TfidfVectorizer(min_df = 5).fit(X_train)
len(vect.get_feature_names())

#we have words which appear across all the reviews

12552

In [28]:
X_train_vectorized = vect.transform(X_train)
model1 = LogisticRegression()
model1.fit(X_train_vectorized, y_train)

predictions = model1.predict(vect.transform(X_test))
print('AUC: ', roc_auc_score(y_test, predictions))

C:\Users\jyost\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


AUC:  0.7784422075995685


##### Number of features(words) reduced from 26280 to 12552 and the AUC score is reduced

In [29]:
feature_names = np.array(vect.get_feature_names())
sorted_tfidf_index = X_train_vectorized.max(0).toarray()[0].argsort()

print('Smallest Tfidf: \n{}\n'.format(feature_names[sorted_tfidf_index[:10]]))
print('Largest Tfidf: \n{}\n'.format(feature_names[sorted_tfidf_index[:-11:-1]]))

Smallest Tfidf: 
['1967' 'clicking' 'stern' 'citizens' 'correlated' 'seventy' 'connected'
 'gala' 'abide' '1940s']

Largest Tfidf: 
['yumm' 'el' 'br' 'bisquick' 'excellent' 'gm' 'thank' 'mold' 'ginger'
 'nom']



### Testing the model

In [31]:
print(model1.predict(vect.transform(['The quality of the food is not good, I will never buy them again','The taste is not bad, I will buy them again'])))

[1 0]


In [41]:
# Text 1: "The quality of the choco bar is not good, I will never buy them again" --> Taken as a positive review 
# Text 2: "The taste is not bad, I will buy them again" --> Taken as a negative review

### n-grams: One way to fix this misclassification

In [32]:
vect = CountVectorizer(min_df = 5, ngram_range = (1,2)).fit(X_train)
X_train_vectorized = vect.transform(X_train)
len(vect.get_feature_names())

127492

In [33]:
model1 = LogisticRegression()
model1.fit(X_train_vectorized, y_train)
predictions = model1.predict(vect.transform(X_test))
print('AUC: ', roc_auc_score(y_test, predictions))
#more words are more AUC Score

C:\Users\jyost\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


AUC:  0.9476229258474729


##### Using the coefficients to check each feature


In [34]:
feature_names = np.array(vect.get_feature_names())
sorted_coef_index = model1.coef_[0].argsort()
print('Smallest Coef: \n{}\n'.format(feature_names[sorted_coef_index][:10]))
print('Largest Coef: \n{}\n'.format(feature_names[sorted_coef_index][:-11:-1]))

Smallest Coef: 
['terrible' 'gross' 'awful' 'horrible' 'disappointing' 'disappointed'
 'yuck' 'nasty' 'at all' 'not very']

Largest Coef: 
['delicious' 'not too' 'smooth' 'nice' 'great' 'fantastic' 'not bitter'
 'perfect' 'not bad' 'excellent']



### Testing the model

In [35]:
print(model1.predict(vect.transform(['The quality of the food is not good, I will never buy them again','The taste is not bad, I will buy them again'])))

[0 1]


### Multinomial Naive Bayes Algorithm - Model 2

In [36]:
from sklearn.naive_bayes import MultinomialNB

In [37]:
model2 = MultinomialNB()
model2.fit(X_train_vectorized,y_train)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [38]:
from sklearn.metrics import roc_auc_score

In [39]:
predictions = model2.predict(vect.transform(X_test))
print('AUC: ', roc_auc_score(y_test,predictions))

AUC:  0.9209762980270613


In [40]:
print(model2.predict(vect.transform(['The quality of the food is not good, I will never buy them again','The taste is not bad, I will buy them again'])))

[0 1]
